[v6r16] Monitoring the space left on DIRAC SEs

vipersec opened the pull request:
I have added a new command named `FreeDiskSpace` which is used to get the total and the remaining disk space of all dirac storage elements that are found in the CS and inserts the results in the `SpaceTokenOccupancyCache` table of `ResourceManagementDB` database.

chaen commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -50,6 +50,48 @@
+    free = (st.f_bavail * st.f_frsize) / convert
+    return S_OK( round(free, 2) )
+
+def totalSpace(path, size = 'TB'):
```
You can most certainly factorize diskSpace and totalSpace.

chaen commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -176,6 +218,20 @@ def export_getMetadata( self, fileID ):
     """
     return self.__getFileStat( self.__resolveFileID( fileID ) )
 
+  types_getRemainingDiskSpace = [StringTypes]
```
Replace ```StringTypes``` with ```basestring```

chaen commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -176,6 +218,20 @@ def export_getMetadata( self, fileID ):
+    return diskSpace(path, size)
+
+  types_getTotalDiskSpace = [StringTypes]
+  def export_getTotalDiskSpace( self, path, size = 'TB' ):
```
Replace ```StringTypes``` with ```basestring```

chaen commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -50,6 +50,48 @@
+    else:
+      return S_ERROR( "No valid size specified" )
+
+    st = os.statvfs(path)
```
Exceptions might be thrown no ?

vipersec replied:
I added this https://github.com/DIRACGrid/DIRAC/pull/3099/commits/7892d014b94b2bc39871d9e20a36d2ec06043f09 Is it enough? Or should I use a Try-Catch statement?

chaen replied:
you need a try catch for all possible exception that might arrise. I am pretty sure there are OSError if the path does not exist for example

chaen commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -176,6 +218,20 @@ def export_getMetadata( self, fileID ):
     return self.__getFileStat( self.__resolveFileID( fileID ) )
 
+  types_getRemainingDiskSpace = [StringTypes]
+  def export_getRemainingDiskSpace( self, path, size = 'TB' ):
```
Detail, but 'free' is probably mroe common than 'remaining'

chaen commented on ResourceStatusSystem/Command/RemainingDiskSpaceCommand.py:
```diff
@@ -0,0 +1,102 @@
+__RCSID__ = '$Id:  $'
+
+
+class RemainingDiskSpaceCommand( Command ):
```
Same general remark concerning 'free' rather than 'remaining'

chaen commented on ResourceStatusSystem/Command/RemainingDiskSpaceCommand.py:
```diff
@@ -0,0 +1,102 @@
+
+    result = ""
+    for attribute in attributes:
+      res = gConfig.getValue( "/Resources/StorageElements/%s/AccessProtocol.1/%s" % ( SE, attribute ) )
```
This is not good because you make assumption as of the section name ```AccessProtocol.1```.  Maybe you should rather use the storageElementFactory rather than gConfig to get what you want.

chaen commented on ResourceStatusSystem/Command/RemainingDiskSpaceCommand.py:
```diff
@@ -0,0 +1,102 @@
+    self.rpc = None
+    self.rsClient = None
+
+  def getUrl(self, SE, protocol = None ):
```
instantiate the SE and call getBasePath (or something very simmilar, I don't remember exactly)

chaen commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -50,6 +50,39 @@
+        # return free space
+        result = (st.f_bavail * st.f_frsize) / convert
+
+    except Exception as e:
```
check the doc, but I think it is enough to catch OSError

vipersec replied:
`Exception` catches all exceptions including `OSError` as far as I know :)

chaen replied:
of course it does, but it is a bad practice to use a catch all if you can have a finer grain selection :) Pylint would tell you as well ;-)

atsareg commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -50,6 +50,39 @@
 MAX_STORAGE_SIZE = 0
 USE_TOKENS = False
 
+def diskSpace(path, size = 'TB', total = False):
```
According to DIRAC coding conventions, method name should start with a verb, e.g. getDiskSpace()

chaen commented on ResourceStatusSystem/Command/FreeDiskSpaceCommand.py:
```diff
@@ -0,0 +1,102 @@
+    self.rpc = None
+    self.rsClient = None
+
+  def getUrl(self, SE, protocol = None ):
```
Seems like my previous comment was lost, but please don't use directly gConfig for that, use the StorageFactory, and don't construct the URLs yourself but use the getBaseURL method of the SE plugins

chaen replied:
Once  merged https://github.com/DIRACGrid/DIRAC/pull/3123

```
In [4]: se = StorageElement(SE)
In [5]: se.getStorageParameters(protocol = protocol)
{'OK': True,
 'Value': {'Host': 'lhcb-se-dirac.cern.ch',
  'Path': '/DataManagement/StorageElement',
  'PluginName': 'DIP',
  'Port': '9148',
  'Protocol': 'dips',
  'SpaceToken': '',
  'StorageName': 'LogSE',
  'URLBase': 'dips://lhcb-se-dirac.cern.ch:9148/DataManagement/StorageElement',
  'WSUrl': ''}}
```

fstagni commented on ResourceStatusSystem/Command/FreeDiskSpaceCommand.py:
```diff
@@ -0,0 +1,102 @@
+
+    return urls
+
+  def doCommand( self ):
```
Hi Steve,
this is not the way you code a command in DIRAC RSS: 
- a command should work on one element only
- this element is in the `params`
- the command should then be called from the CacheFeederAgent

fstagni replied:
And you should provide a `doNew`, `doMaster`, `doCache` methods

chaen replied:
Just in case there is no, it would be a good opportunity to write a doc about it.

vipersec commented:
So, in https://github.com/DIRACGrid/DIRAC/pull/3099/commits/fac178313573f3cd58d81d47b437b00e2ccdbcbb I added `doNew`, `doMaster`, `doCache` methods and changed `doNew` to work on one element only.

Is it ok to have a `for` loop to get all the storage element names in `doMaster`?

fstagni commented on ResourceStatusSystem/Command/FreeDiskSpaceCommand.py:
```diff
@@ -0,0 +1,115 @@
+
+    result = ""
+    for attribute in attributes:
+      res = gConfig.getValue( "/Resources/StorageElements/%s/AccessProtocol.1/%s" % ( SE, attribute ) )
```
Please follow the advice that Chris gave you few lines above

fstagni commented on ResourceStatusSystem/Command/FreeDiskSpaceCommand.py:
```diff
@@ -0,0 +1,115 @@
+
+    if 'element' not in self.args:
+      return S_ERROR( '"element" not found in self.args' )
+    element = self.args[ 'element' ]
```
`element` is in RSS usally either "Site" or "Resource". I think what you are looking for here is `name` (the name of the resource)

fstagni commented on ResourceStatusSystem/Command/FreeDiskSpaceCommand.py:
```diff
@@ -0,0 +1,115 @@
+
+  def doNew( self, masterParams = None ):
+    """
+    Gets the total and the free disk space of all DIPS storage elements that
```
nope, just one

fstagni commented on ResourceStatusSystem/Command/FreeDiskSpaceCommand.py:
```diff
@@ -0,0 +1,115 @@
+
+    return S_OK()
+
+  def doCache( self ):
```
This should be implemented to get it from the spaceTokenOccupancy cache.

fstagni commented on ResourceStatusSystem/Command/FreeDiskSpaceCommand.py:
```diff
@@ -0,0 +1,115 @@
+  def doCache( self ):
+    return S_OK()
+
+  def doMaster( self ):
```
add a docstring

fstagni commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -50,6 +50,39 @@
+        result = (st.f_bavail * st.f_frsize) / convert
+
+    except OSError as e:
+      return S_ERROR( "Error %s", e )
```
do instead
```
return S_ERROR( errno.EIO, "Error while getting the available disk space: %s", repr(e) )
```

fstagni commented:
BTW, it is ok to have a for loop to get all the storage element names in `doMaster`. 

And please add some doc.

fstagni commented on ResourceStatusSystem/Command/FreeDiskSpaceCommand.py:
```diff
@@ -95,15 +95,34 @@ def doNew( self, masterParams = None ):
+    result = self.rsClient.addOrModifySpaceTokenOccupancyCache(endpoint = elementURL, lastCheckTime = datetime.utcnow(),
+                                                               free = free, total = total, token = elementName )
+    if not result[ 'OK' ]:
+      return S_ERROR( DErrno.EMYSQL, "Query failed %s" % result )
```
`result['Message']`

chaen replied:
no,
```
if not result['OK']:
    return result
```
Let the error propagate

chaen commented on ResourceStatusSystem/Command/FreeDiskSpaceCommand.py:
```diff
@@ -0,0 +1,134 @@
+    result = self.rsClient.selectSpaceTokenOccupancyCache(token = elementName)
+
+    if not result[ 'OK' ]:
+      return S_ERROR( DErrno.EMYSQL, "Query failed %s" % result['Message'] )
```
Same here

JanEbbing commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -50,6 +51,39 @@
+      If total is set to true, the total disk space will be returned instead.
+    """
+
+    if size.upper() == "KB":
```
I would replace this with something like
```python
UNIT_CONVERSION = { "KB" : 1024, "MB" : 1.048576e6, ... }
def getDiskSpace(...):
  size_to_convert = size.upper()
  if size_to_convert not in UNIT_CONVERSION:
    return S_ERROR( "no valid size" )
  convert = UNIT_CONVERSION[size_to_convert]
```

This is shorter, imo easier to understand, debug/maintain and in general better coding style. There's probably even better solutions (I don't like the total flag) but that should be ok.

phicharp replied:
and it is also much better to not define the units in floating point but as integers!

MB: 1024*1024
GB: 1024*1024*1024
etc...

vipersec replied:
Done and done :)

JanEbbing commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -50,6 +51,39 @@
+    try:
+      st = os.statvfs(path)
+
+      if total:
```
Also here, slightly better imo:
```python
if total:
  queried_size = st.f_blocks
else:
  queried_size = st.f_bavail
result =  ( queried_size * st.f_frsize ) / convert
```

JanEbbing commented on ResourceStatusSystem/Command/FreeDiskSpaceCommand.py:
```diff
@@ -0,0 +1,102 @@
+      return S_OK()
+
+    self.rpc = RPCClient( elementURL, timeout=120 )
+    free = self.rpc.getFreeDiskSpace("/")
```
Might be (probably is) premature optimization, but do these queries take long?
You could split `getDiskSpace` even more so that you have (private) methods that take the result of a `os.statvfs` and return the free or the total space of a disk. Then 3 public methods `getFreeDiskSpace`, `getTotalDiskSpace` and `getDiskInfo` which returns a tuple ( freeSpace, totalSpace ) (would violate SRP but I could imagine a `os.stat('/')` to run quite a while).

vipersec replied:
The queries do not take long normally as far as I know. I will stick to the current solution for the time being.

phicharp commented:
Not related to this PR: I find the new way of commenting a PR absolutely tedious... It is not clear at all when a notification is sent to the participants, and since notifications are not sent to the author of it there is not way to check...
Anyway probably useless to say that here as nobody will be notified...
Can you please confirm you received something? ;-)

phicharp commented:
@vipersec : although approved can you take Jan's comments into account (and mine occasionally)

sposs commented:
I do receive all comments.

Le jeu. 29 sept. 2016 à 13:19, Philippe Charpentier <
notifications@github.com> a écrit :

> @vipersec <https://github.com/vipersec> : although approved can you take
> Jan's comments into account (and mine occasionally)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DIRACGrid/DIRAC/pull/3099#issuecomment-250438854>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAvHo3BAotrHWXm8AucUj8O5ajpa8mr7ks5qu56kgaJpZM4J09Ix>
> .
>

andresailer commented on DataManagementSystem/Service/StorageElementHandler.py:
```diff
@@ -50,6 +51,39 @@
+        result = (st.f_bavail * st.f_frsize) / convert
+
+    except OSError as e:
+      return S_ERROR( errno.EIO, "Error while getting the available disk space: %s", repr(e) )
```
There should be a "%" instead of the comma after the second argument.
